compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/cfa/CfaUtils.kt

```diff
@@ -60,21 +59,59 @@ class PropertyInitializationInfo(
         ::EMPTY
 }
 
-class LocalPropertyCollector private constructor() : ControlFlowGraphVisitorVoid() {
+class LocalPropertyAndCapturedWriteCollector private constructor() : ControlFlowGraphVisitorVoid() {
     companion object {
-        fun collect(graph: ControlFlowGraph): MutableSet<FirPropertySymbol> {
-            val collector = LocalPropertyCollector()
+        fun collect(graph: ControlFlowGraph): Pair<Set<FirPropertySymbol>, Set<FirVariableAssignment>> {
+            val collector = LocalPropertyAndCapturedWriteCollector()
             graph.traverse(TraverseDirection.Forward, collector)
-            return collector.symbols
+            return collector.symbols.keys to collector.capturedWrites
         }
     }
 
-    private val symbols: MutableSet<FirPropertySymbol> = mutableSetOf()
+    // Mapping from a property symbol to its declaration context
+    // `null` if the (local) property is declared in the currently visited function.
+    private val symbols: MutableMap<FirPropertySymbol, FirFunction<*>?> = mutableMapOf()
+
+    private val lambdaOrLocalFunctionStack: MutableList<FirFunction<*>> = mutableListOf()
+    private val capturedWrites: MutableSet<FirVariableAssignment> = mutableSetOf()
 
     override fun visitNode(node: CFGNode<*>) {}
 
     override fun visitVariableDeclarationNode(node: VariableDeclarationNode) {
-        symbols += node.fir.symbol
+        symbols[node.fir.symbol] = lambdaOrLocalFunctionStack.lastOrNull()
+    }
+
+    override fun visitPostponedLambdaEnterNode(node: PostponedLambdaEnterNode) {
+        lambdaOrLocalFunctionStack.add(node.fir)
+    }
+
+    override fun visitPostponedLambdaExitNode(node: PostponedLambdaExitNode) {
+        lambdaOrLocalFunctionStack.remove(node.fir)
+    }
+
+    override fun visitLocalFunctionDeclarationNode(node: LocalFunctionDeclarationNode, data: Nothing?) {
+        lambdaOrLocalFunctionStack.add(node.fir)
+    }
+
+    override fun visitFunctionExitNode(node: FunctionExitNode) {
+        lambdaOrLocalFunctionStack.remove(node.fir)
+    }
+
+    override fun visitVariableAssignmentNode(node: VariableAssignmentNode) {
+        // Check if this assignment is inside a lambda or a local function.
+        if (lambdaOrLocalFunctionStack.isEmpty()) return
+
+        // Check if the assigned variable doesn't belong to any lambda or local function.
+        val symbol = node.fir.referredPropertySymbol ?: return
+        if (symbol !in symbols || symbols[symbol] != null) return
+
+        // If all nested declarations are lambdas that are either inlined or invoked in-place, it's not a captured write.
+        if (lambdaOrLocalFunctionStack.all {
+                it is FirAnonymousFunction && (it.inlineStatus == InlineStatus.Inline || it.invocationKind != null)
```

Comment:
This `...invocationKind != null` means it's invoked in-place, and there is a (`private`) util: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/ControlFlowGraphBuilder.kt#L309

I'm not sure where is a good place to utilize that. Any suggestions?

Reply:
I think it's OK to move it to some `CfgUtils` in `org.jetbrains.kotlin.fir.resolve.dfa.cfg` package

Reply:
Thanks for the suggestion. Added.